src/string_id.h

```diff
@@ -65,11 +71,115 @@ class generic_factory;
  *      conversion to int_id is extremely fast (just returns int field)
  *     `::obj` call is relatively fast (array read by int index), a bit slower than on int_id
  */
+
+/** @relates string_id_params */
+template<bool IsDynamic>
+struct abstract_string_id_params {
+    static constexpr bool dynamic = IsDynamic;
+};
+/**
+ *  Type marker that specifies whether string ids of a specific type
+ *  are "static" (i.e. "interned", having a single pointer inside) or
+ *  "dynamic" (non-interned, having std::string inside).
+ *
+ *  By default all string ids are static/interned.
+ *  To make a specific string_id type dynamic, add the following specialization:
+ *      template<> struct string_id_params<Trait_group>: abstract_string_id_params<true> {};
+ *  Replace `T` above with concrete type of the string id  (see type_id.h for examples)
+ */
+template<typename T>
+struct string_id_params: abstract_string_id_params<false> {};
+
+// mark string_ids of certain classes as "dynamic"
+// NOTE: string_id.h is the safest place for these markers, because they MUST be declared
+//  before any usage of corresponding string_id type in cpp files
+class Item_spawn_data;
+template<> struct string_id_params<Item_spawn_data>: abstract_string_id_params<true> {};
+class Trait_group;
+template<> struct string_id_params<Trait_group>: abstract_string_id_params<true> {};
+
+// Two different implementations of the "identity" part of string_id.
+/**
+ * "static" here means std::string inside is interned.
+ * Can be used only if all string_id values are fixed in number and not generated dynamically.
```

Comment:
That's definitely something that I want to look into at some point. I have some ideas about other checks as well (like, using int_ids in static context or constructing non-static string ids inline). However, I won't make any promises as to when I get to do that, as I have some other PRs that I want to finish first. 

So feel free to add this check yourself if you want/have time, otherwise I may do that at some point in the future.